### PR TITLE
[CRT] Improve rand_s

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

This improves rand_s to avoid the massive overhead of loading advapi32, opening a handle to ksecdd, closing the handle and unloading advapi32 for every single call to rand_s. Instead the function pointer is cached. Testing resulted in a speedup of 10,000 times when run on Windows 10 using the statically linked apitest.
This also adds rand_s to msvcrtex for linking with newer GCC stdc++.

## Tests
- Test WHS: https://reactos.org/testman/compare.php?ids=94813,94812 (static_crt:rand_s)
before: 5.94s - `rand_s took 13040596278 cycles`
after: 0.02s - `rand_s took 11622704 cycles`
=> 1000x

- KVM x86: https://reactos.org/testman/compare.php?ids=94818,94824
before: 0.51s - `rand_s took 1070359452 cycles`
after: 0.15s  `rand_s took 285251108 cycles`
=> 3.7x

This is a quite unexpected result, showing that our totally broken and in no way cryptographically secure user mode implementation of RtlGenRandom is actually 25x slower than the kernel mode based one in win 2003, while the overhead of loading advapi32 and opening the handle to ksecdd is somehow 12 times faster on reactos.

### CODE REVIEW

File: drivers/crypto/ksecdd/client/ksecclient.c [hunk 1]
```diff
@@ -0,0 +1,94 @@
+
```
Comment (HBelusca) [inline drivers/crypto/ksecdd/client/ksecclient.c:1]: Suggestion: Add standard file header with copyright and license.

File: drivers/crypto/ksecdd/client/ksecclient.c [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+NTSTATUS
+NTAPI
+KsecDeviceIoControl(
+    ULONG IoControlCode,
+    PVOID InputBuffer,
+    SIZE_T InputBufferLength,
+    PVOID OutputBuffer,
+    SIZE_T OutputBufferLength)
```
Comment (HBelusca) [inline drivers/crypto/ksecdd/client/ksecclient.c:64]: SAL2 annotations? (OK they are already in header, so either you copy-paste them here too, or use `_Use_decl_annotations_` at the top of the function.

File: drivers/crypto/ksecdd/client/ksecclient.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
+
```
Comment (HBelusca) [inline drivers/crypto/ksecdd/client/ksecclient.h:1]: File header + copyright/license.

File: sdk/lib/crt/stdlib/_invalid_parameter.c [hunk 4]
```diff
@@ -0,0 +1,52 @@
+/*
+ * PROJECT:     ReactOS CRT library
+ * LICENSE:     MIT (https://spdx.org/licenses/MIT)
+ * PURPOSE:     _invalid_parameter implementation
+ * COPYRIGHT:   Timo Kreuzer <timo.kreuzer@reactos.org>
```
Comment (HBelusca) [inline sdk/lib/crt/stdlib/_invalid_parameter.c:5]: Usually we put `Copyright (the_year) Author <email>`

File: sdk/lib/crt/stdlib/stdlib.cmake [hunk 5]
```diff
@@ -1,11 +1,13 @@
 
 list(APPEND LIBCNTPR_STDLIB_SOURCE
+    stdlib/rand_nt.c
     stdlib/qsort.c
```
Comment (HBelusca) [inline sdk/lib/crt/stdlib/stdlib.cmake:4]: qsort should be before rand_nt when sorting the file list in alphabetical order.

File: drivers/crypto/ksecdd/client/ksecclient.c [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    NTSTATUS Status;
+
+    /* Check if we already have a handle */
+    if (KsecDeviceHandle == NULL)
+    {
+        /* Try to open the device */
+        Status = KsecOpenDevice();
+        if (!NT_SUCCESS(Status))
+        {
+            //ERR("Failed to open handle to KsecDd driver!\n");
```
Comment (HBelusca) [inline drivers/crypto/ksecdd/client/ksecclient.c:76]: DPRINT or DPRINT1 instead?

**Review by HBelusca — APPROVED** [review]
OK, just some minor suggestions here.
And 3rd commit: "its own file", not "it's".

File: drivers/crypto/ksecdd/client/ksecclient.c (hunk 6 above)
Comment (JoachimHenze) [inline drivers/crypto/ksecdd/client/ksecclient.c:76]:
```suggestion
            //ERR("Failed to open handle to KsecDd driver\n");
```

File: drivers/crypto/ksecdd/client/ksecclient.h (hunk 3 above)
Comment (JoachimHenze) [inline drivers/crypto/ksecdd/client/ksecclient.h:1]: Where is the #pragma once? (or at least an include guard)

File: sdk/lib/crt/stdlib/rand_s.c [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * PROJECT:     ReactOS CRT library
+ * LICENSE:     MIT (https://spdx.org/licenses/MIT)
+ * PURPOSE:     rand_s implementation
+ * COPYRIGHT:   Timo Kreuzer <timo.kreuzer@reactos.org>
+ */
+
+#include <precomp.h>
+#include <ksecclient.h>
+
+/*!
```
Comment (JoachimHenze) [inline sdk/lib/crt/stdlib/rand_s.c:11]:
```suggestion
/*
```
Assuming that the exclamation mark has no meaning, but was a typo.

**Comment by JoachimHenze** [comment]
Did you have any specific testcase, which made you aware of the overhead? (e.g. by poor performance in a specific app?)

File: sdk/lib/crt/stdlib/rand_s.c (hunk 7 above)
Comment (learn-more) [reply sdk/lib/crt/stdlib/rand_s.c:11]: If you mean 'doxygen' with 'no meaning' then yes.

**Comment by tkreuzer** [comment]
> Did you have any specific testcase, which made you aware of the overhead? (e.g. by poor performance in a specific app?)

I wrote a test and that showed that the new implementations isn't "compatible" with Windows, so you cannot run it as a unittest. I decided to instead use the previous code, but instead of loading and unloading advapi32 for every call, the function pointer is cached. This speeds up the statically linked performance test by a factor of 10,000 on Windows 10. Our testbots will show what the benefit is on windows 2003 / reactos.

**Comment by RaulKong898** [comment]
Hi. I would like ReactOS to look and feel more like Windows XP, especially the system apps to be more like Windows XP. I also have the Windows XP source code that was leaked on the internet. This is the source code link: https://github.com/tongzx/nt5src. I really want ReactOS to look like Windows XP, especially the system apps, because Windows XP is no longer supported and we can easily get viruses and stuff. So with ReactOS we can modify the source code of Windows XP and add interesting things that never existed on Windows XP and we can have security support. Consider this message, especially since many people who lived with Windows XP would like to go back to Windows XP because it seemed simpler and different, especially for retro tech people and for older PCs that they are no longer used and if they are thrown away we destroy the planet if they are not recycled, especially since this way ReactOS will have support for many drivers from Windows XP.

**Pull request merged**